hibernate-core/src/main/java/org/hibernate/engine/internal/StatefulPersistenceContext.java

```diff
@@ -489,13 +489,29 @@ public EntityEntry addEntry(
 
 		final EntityEntry e;
 
-		if( (entity instanceof ManagedEntity) &&  ((ManagedEntity) entity).$$_hibernate_getEntityEntry() != null && status == Status.READ_ONLY) {
-			e = ((ManagedEntity) entity).$$_hibernate_getEntityEntry();
-			e.setStatus( status  );
+		/*
+		Important: 		the following instanceof checks and castings are intentional.
+						Please do *not* refactor to make calls through the EntityEntryFactory
+						interface. Refactoring will results in polymorphic call sites which will
+						severely impact performance.
+		*/
+		if (persister.getEntityEntryFactory() instanceof MutableEntityEntryFactory){
+			e = ((MutableEntityEntryFactory) persister.getEntityEntryFactory()).createEntityEntry(
+					status,
+					loadedState,
+					rowId,
+					id,
+					version,
+					lockMode,
+					existsInDatabase,
+					persister,
+					disableVersionIncrement,
+					lazyPropertiesAreUnfetched,
+					this
+			);
 		}
 		else {
-			final EntityEntryFactory entityEntryFactory = persister.getEntityEntryFactory();
-			e = entityEntryFactory.createEntityEntry(
+			e = ((ImmutableEntityEntryFactory) persister.getEntityEntryFactory()).createEntityEntry(
 					status,
```

Comment:
With the patch, there is a type check and a non-local jump
Without the patch, this is a type check, 3 memory references and 2 non local jumps
The JVM can not cache the vtable resolution as there is more than one class loaded that implements the virtual method, and needs to resolve this everytime.  If it were to cache the resolution, you would have method calls being dispatched to the wrong class.

When there is only 1 implementation of the virtual method, the JVM knows that there can only be one concerete method, so replaces the virtual method call with a direct call and removes the vtable lookup.  This was the case in ORM4.2.  When we introduced another implementation of the EntityEntryFactory interface, the JVM can no longer make this optimisation.

Another effect is, in 4.2, the JVM could substitute the virtual method call for a direct method call, then it was able to inline that method at call sites.  With 2 concrete methods, the JVM is unable to inline the method calls at the call sites as it needs to resolve which implemenation to use